### ISSUE TITLE
Use async methods all the way

### ISSUE THREAD

**Issue by pohmelie** [eerimoq/asyncudp#1]
I think you should mimic the asyncio «way» and mark all methods as `async`. Since there is no profit of using some methods as `async def` and some as a `def`.

**Comment by eerimoq** [comment]
That is far from true for many standard library modules. Have a look at TCP sockets for example: https://docs.python.org/3/library/asyncio-stream.html#streamwriter

I'm not saying that a mix of `async def` and `def` is good, just that mixing them is common.

**Comment by pohmelie** [comment]
`asyncio` interface is famous «first experience» with much architecture problems. For example, duality of `StreamWriter.write` and `StreamWriter.drain`. If you look at modern asynchornous io frameworks for python like [`trio`](https://trio.readthedocs.io/en/stable/reference-io.html#low-level-networking-with-trio-socket), you will see all methods are async. So this is very rare case where you really need or prefer non-async methods.
This also leads to problem, where user can «write» some data with non-async call in a loop and eat all memory.

**Comment by XChikuX** [comment]
@pohmelie Low CPU and low resource usage `def` is fine. High resource usage (Eg. disk, memory, network, etc) `async def` is a must.

This is because you want to minimize event loop blockages.
There is a tiny overhead to awaiting every function though. 

So if you have just a couple of lines of code.  `def` usage is perfectly fine.

I have verified this with industry veterans.

**Comment by pohmelie** [comment]
Sorry for a late answer. Try this code and checkout the memory consumption:
```py
import asyncio
from itertools import count


async def server_handler(reader, writer):
    while True:
        data = await reader.read(2 * 10)
        print(f"received {len(data) = }")


async def main():
    server = await asyncio.start_server(server_handler, "127.0.0.1", 8080)
    async with server:
        reader, writer = await asyncio.open_connection("127.0.0.1", 8080)
        for _ in count():
            data = b"0" * 2 ** 20
            writer.write(data)
            # await asyncio.sleep(1)

asyncio.run(main())
```
This is what I'm talking about.

**Comment by pohmelie** [comment]
Also, you can see stuck on this code:
```py
import asyncio
import asyncudp

async def client():
    sock = await asyncudp.create_socket(remote_addr=("127.0.0.1", 9999))
    while True:
        data = b"0" * 2 ** 10
        sock.sendto(data)

async def server():
    sock = await asyncudp.create_socket(local_addr=("127.0.0.1", 9999))
    while True:
        data, addr = await sock.recvfrom()
        print(f"{addr = } {len(data) = }")

async def main():
    async with asyncio.TaskGroup() as tg:
        tg.create_task(server())
        tg.create_task(client())

asyncio.run(main())
```

**Comment by arthur-tacca** [comment]
> If you look at modern asynchornous io frameworks for python like trio, you will see all methods are async.

That is not true. In Trio, methods are async if it makes sense for them to be, and they are not if it doesn't. For example, `CancelScope.cancel()` and `Nursery.start_soon()` are both non-async.

The reasoning cuts both ways: If a function cannot block (in the async sense) then it's useful to be able to call it from a non-async connect. Conversely, when a library follows this time and you see a function is async then you know it might need to do some I/O that could block (in the async sense).

(Trio actually takes this further by guaranteeing that all Trio async functions give other tasks the opportunity to run even if the current function could complete synchronously, e.g., when you read from a socket that already has some data waiting. So it makes even less sense to just mark everything as an asynchronous function in Trio.)

**Comment by arthur-tacca** [comment]
Ah, looking at the later comments, I see that what I said isn't really relevant, sorry.

As I said, Trio does use regular sync methods when it's appropriate. But, for the specific case of writing to a socket (including UDP), it does have a single async `write()` method, rather than asyncio's sync `write()` + async `drain()` pairing. That's because Trio's `write()` directly issues a system call and waits for it to complete (though of course that doesn't mean the data has reached the destination), whereas asyncio has its own weird buffer on top of the OS one. Trio's creator, NJS, wrote [a good blog post about why this is a bad idea](https://vorpus.org/blog/some-thoughts-on-asynchronous-api-design-in-a-post-asyncawait-world/#bug-1-backpressure).

asyncudp doesn't follow either approach: with asyncio you have to remember to call `drain()`, but asyncudp doesn't even make `drain()` available. How do you know when the call to the OS has completed?
